### PR TITLE
Unhandled Exception for malformed connection string bug fix

### PR DESCRIPTION
**Pull request by aaronburtle**
## Why make this change?

Closes https://github.com/Azure/data-api-builder/issues/1736

When a connection string was not properly formed, in the case of `MSSQL` database types it was possible to have an exception when modifying the connection string as part of our normal code. However, that exception was not being handled and would cause an ungraceful crash of the application. We now handle this exception.

## What is this change?

Include `DataApiBuilderException` in the exception handling for where we modify the connection string in this particular case. And include a regression test case.

New error messaging can be seen here:

image

It should also be noted that some tests had expected behavior that did not account for our new exception handling. Some test cases have been changed to `postgresql` data base type to avoid conflicts with our new post processing error handling, when that new error handling is not relevant to what those tests were covering. Since the post processing only happens in the following conditional, this is an easy way to keep our test coverage without major changes.

image

Our tests for error checking bad connection strings for mssql was also relying on having the actual exception available, but we now have exception handling that will catch that and return a different exception than the test is expecting. To keep this test working we redirect standard error and then checking against that messaging for those cases where we are both mssql db type and have a malformed connection string . The remaining difference for those cases that do not have valid connection strings and are mssql db type is that we have an `InternalServerError` as the status code now, so we modify the test to look for that instead in this case.

## How was this tested?

A regression test is added to the unit tests for the runtime config loader. This is where the change happened, and the new case covers a malformed connection string, verifying that we handle the exception correctly by validating that we return false from the function that was throwing the unhandled exception.

## Sample Request(s)

You can manually repro with the instructions from the linked issue this closes which are the following steps:

1. After building the solution, from the root of the repo: cd src\out\cli\net6.0
2. Microsoft.DataApiBuilder.exe init --database-type mssql --connection-string "abcd"
3. Microsoft.DataApiBuilder.exe add Book --source dbo.books --permissions "anonymous:*"
4. Microsoft.DataApiBuilder.exe start

### CODE REVIEW

File: src/Config/RuntimeConfigLoader.cs [hunk 1]
```diff
@@ -145,7 +145,9 @@ public RuntimeConfigLoader(string? connectionString = null)
             }
 
         }
-        catch (JsonException ex)
+        catch (Exception ex) when (
+            ex is JsonException ||
+            ex is DataApiBuilderException)
```
Comment (Aniruddh25) [inline src/Config/RuntimeConfigLoader.cs:150]: if the exception is DataApiBuilderException, is the error message here still a valid message? How is the deserialization failure message valid for a connection string failure?

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
Thanks for the quick fix and test but needs better messaging to convey exactly what is wrong with the configuration so the users can self help themselves.

File: src/Config/RuntimeConfigLoader.cs (hunk 1 above)
Comment (aaronburtle) [reply src/Config/RuntimeConfigLoader.cs:150]: I thought it was still descriptive for what is happening because this is a post processing step of the deserialization. We deserialize directly from the json string to create the objects, then we have a conditional update as part of the post processing. So in my eyes that was still part of the deserialization process and giving it another name in the error messaging might be confusing.

I considered calling this "parsing" instead but thought it was clearer like this. this is what it looks like now

image

File: src/Config/RuntimeConfigLoader.cs [hunk 2]
```diff
@@ -145,7 +145,9 @@ public RuntimeConfigLoader(string? connectionString = null)
             }
 
         }
-        catch (JsonException ex)
+        catch (Exception ex) when (
+            ex is JsonException ||
```
Comment (seantleonard) [inline src/Config/RuntimeConfigLoader.cs:149]: I see what Ani is saying. JsonException directly correlates to a deserialization issue. However, the DAB exception could be something else, like in this case with a failure to build a connection string.
Perhaps a separate catch for the DAB exception and log the associated error message.

Comment (aaronburtle) [reply src/Config/RuntimeConfigLoader.cs:149]: I am just not sure that differentiating between the post processing done during deserialization and the core deserialization itself is going to be a key distinction we want to make in the error messaging.

Comment (aaronburtle) [reply src/Config/RuntimeConfigLoader.cs:149]: As an example of what I mean, suppose that the Json Deserializer added a post processing option where you could pass in a function and it would execute that function as a post processing step for you.

from the users perspective, the same thing is happening, would we differentiate that exception? 

this seems like a detail that is really only visible at the implementation level, unless we don't consider the post processing steps part of the deserialization process generally, which I do.

File: src/Cli.Tests/EnvironmentTests.cs [hunk 3]
```diff
@@ -162,8 +164,7 @@ public async Task FailureToStartEngineWhenEnvVarNamedWrong()
         );
 
         string? output = await process.StandardError.ReadLineAsync();
-        StringAssert.Contains(output, "Environmental Variable, "
-            + expectedEnvVarName + ", not found.", StringComparison.Ordinal);
+        Assert.AreEqual(output, "Deserialization of the configuration file failed.");
```
Comment (seantleonard) [inline src/Cli.Tests/EnvironmentTests.cs:167]: The previous check explicitly checked for the failure that the test was trying to trigger. Can you include both your new addition to check the high level deserialization failure notice as well as checking `StringAssert.Contains(output, "Environmental Variable, "`?

Also, the test name indicates that this is testing an invalid env_var name provided in config, as opposed to a invalid connection string failing post-processing. Can you clarify test intent here?

File: src/Config/RuntimeConfigLoader.cs (hunk 1 above)
Comment (seantleonard) [reply src/Config/RuntimeConfigLoader.cs:150]: include that screenshot in the pr description to demonstrate this PR's affect.

File: src/Config/RuntimeConfigLoader.cs (hunk 2 above)
Comment (seantleonard) [reply src/Config/RuntimeConfigLoader.cs:149]: I see. I think you point then begs the question of whether the error message we currently set is too specific?

> Deserialization of the configuration file failed.

Since that message specifically distinguishes between deserialization and post-processing, a still relevant but less granular message that wouldn't affect how a user reacts could be:

> Configuration file processing failed.

File: src/Cli.Tests/EnvironmentTests.cs (hunk 3 above)
Comment (aaronburtle) [reply src/Cli.Tests/EnvironmentTests.cs:167]: The error message contains whatever we decide to put in the messaging, which right now is, "Deserialization of the configuration file failed.", to add more details than that we would have to add it to the original message.

Comment (seantleonard) [reply src/Cli.Tests/EnvironmentTests.cs:167]: Understood that we have error handling now and set the message "Deserialization of the configuration file failed." My point is that the test can also check the specific error causing the deserialization issue. Those details are printed to the console as shown by your screenshot: 
image

The environment variable error messaging may still be available to check here, perhaps if you continue reading errors printed to the console `process.StandardError.ReadLineAsync();`.

That way, the test can be specific so you know that no other possible issues are causing the test to pass/fail:
- if you're testing env_var issues, check for both the "Deserialization of the configuration file failed." and exception specific messaging printed to the console
- if you're testing connectionstring builder errors, check for both the "Deserialization of the configuration file failed." and exception specific messaging printed to the console

Comment (aaronburtle) [reply src/Cli.Tests/EnvironmentTests.cs]: sure we can read out the entire error message and verify the underlying exception includes the environment variable error.

File: src/Config/RuntimeConfigLoader.cs (hunk 2 above)
Comment (aaronburtle) [reply src/Config/RuntimeConfigLoader.cs:149]: I added some language about this being during a post processing step, but I am still not sure this really helps the end understand better what is wrong.

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs [hunk 4]
```diff
@@ -83,13 +83,13 @@ public class RuntimeConfigLoaderJsonDeserializerTests
...
                     Assert.IsTrue(RuntimeConfigLoader.TryParseConfig(
-                        GetModifiedJsonString(repValues, @"""mssql"""), out expectedConfig, replaceEnvVar: replaceEnvVar),
+                        GetModifiedJsonString(repValues, @"""postgresql"""), out expectedConfig, replaceEnvVar: replaceEnvVar),
                         "Should read the expected config");
                 }
                 else
                 {
                     Assert.IsTrue(RuntimeConfigLoader.TryParseConfig(
-                        GetModifiedJsonString(repKeys, @"""mssql"""), out expectedConfig, replaceEnvVar: replaceEnvVar),
+                        GetModifiedJsonString(repKeys, @"""postgresql"""), out expectedConfig, replaceEnvVar: replaceEnvVar),
```
Comment (seantleonard) [inline src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:92]: why these changes?

Comment (aaronburtle) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:92]: this test doesnt use a valid connection string format, so if we use mssql db type we end up inside the code where we will get an exception now during post processing. we want to avoid that for this test, so using postgressql is an easy way to do that.

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs [hunk 5]
```diff
@@ -156,16 +156,20 @@ public void CheckConfigEnvParsingThrowExceptions(string invalidEnvVarName)
...
     ""data-source"": {
-        ""database-type"": ""notsupporteddb""
-     }
+        ""database-type"": " + dbType + @",
+        ""connection-string"": " + connectionString + @"
+    },
+    ""entities"":{ }
 }";
-            Assert.IsFalse(RuntimeConfigLoader.TryParseConfig(configJson, out RuntimeConfig deserializedConfig));
+            Assert.IsFalse(RuntimeConfigLoader.TryParseConfig(configJson, out RuntimeConfig deserializedConfig, replaceEnvVar: true));
```
Comment (seantleonard) [inline src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:172]: what is replaceEnvVar doing for the test? could yo add a test description/displaynames for the data rows?

File: src/Config/RuntimeConfigLoader.cs (hunk 2 above)
Comment (seantleonard) [reply src/Config/RuntimeConfigLoader.cs:149]: That message was fine. When exceptions are raised their errors are automatically output to the console, so that gives more detail.

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs (hunk 5 above)
Comment (aaronburtle) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:175]: Sure I added a comment. This is just used to get into the portion of the code that does the post-processing, which is what we are testing for the connection string.

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs (hunk 4 above)
Comment (seantleonard) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:92]: followed up on Teams: Our configloader looks for "MSSQL" dbtype to decide whether to scrutinize the connection string and subject it to the connectionstringbuilder. 

This test is evaluating env_Var replacement which may end up creating an intentionally bad connection string, however, the test is only checking for whether env_vars were processed properly by checking whether configloading succeeded. Now that we have connecitonstringbuilder validation in the runtimeconfigloader, the test broke because of a different untargeted failure.

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs (hunk 5 above)
Comment (seantleonard) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:175]: if we don't replace environment variables, the connectionstringBuilder check / exception won't be hit? wouldn't an unreplace env_var be an invalid connection string?

Comment (aaronburtle) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:175]: the connection string builder is called within this call that happens, and this call is only hit for MSSQL with replace environment variables being true
image

**Comment by seantleonard** [comment]
How does this affect the late-bound config scenario?

File: src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs (hunk 5 above)
Comment (seantleonard) [reply src/Service.Tests/Unittests/RuntimeConfigLoaderJsonDeserializerTests.cs:175]: gotcha, add that to the PR description too and/or note that in the test description comment for the test you changed from mssql->pgsql

**Comment by aaronburtle** [comment]
> How does this affect the late-bound config scenario?

the late bound scenario is just calling this function, the TryParse function, with a string that was sent to the configuration controller, rather than one that was read from a file. this will not cause any special behavior, but it will catch unhandled exceptions that otherwise would have happened if this case was triggered in the hosted scenario. so it does provide coverage there, but nothing special.

File: src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs [hunk 6]
```diff
@@ -73,16 +74,25 @@ public async Task CheckNoExceptionForNoForeignKey()
...
-        [DataRow("DO NOT EDIT, look at CONTRIBUTING.md on how to run tests")]
-        [DataRow("")]
-        public async Task CheckExceptionForBadConnectionStringForMsSql(string connectionString)
+        [DataRow(";;;;;fooBarBAZ", true)]
+        [DataRow("!&^%*&$$%#$%@$%#@()", true)]
+        [DataRow("Server=<>;Databases=<>;Persist Security Info=False;Integrated Security=True;MultipleActiveResultSets=False;Connection Timeout=5;", true)]
+        [DataRow("Servers=<>;Database=<>;Persist Security Info=False;Integrated Security=True;MultipleActiveResultSets=False;Connection Timeout=5;", true)]
+        [DataRow("DO NOT EDIT, look at CONTRIBUTING.md on how to run tests", true)]
+        [DataRow("", false)]
+        public async Task CheckExceptionForBadConnectionStringForMsSql(string connectionString, bool invalidConnectionBuilderString)
```
Comment (Aniruddh25) [inline src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:83]:
```suggestion
        public async Task CheckExceptionForBadConnectionStringForMsSql(string connectionString, bool isInvalidConnectionBuilderString)
```

File: src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs [hunk 7]
```diff
@@ -178,9 +188,19 @@ private static async Task CheckExceptionForBadConnectionStringHelperAsync(string
             catch (DataApiBuilderException ex)
             {
                 // use contains to correctly cover db/user unique error messaging
-                Assert.IsTrue(ex.Message.Contains(DataApiBuilderException.CONNECTION_STRING_ERROR_MESSAGE));
-                Assert.AreEqual(HttpStatusCode.ServiceUnavailable, ex.StatusCode);
+                // if sw is not null it holds the error messaging
+                string error = sw is null ? ex.Message : sw.ToString();
+                Assert.IsTrue(error.Contains(DataApiBuilderException.CONNECTION_STRING_ERROR_MESSAGE));
                 Assert.AreEqual(DataApiBuilderException.SubStatusCodes.ErrorInInitialization, ex.SubStatusCode);
+                if (sw is not null)
```
Comment (Aniruddh25) [inline src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:195]: Check for the newly added post processing error message here in the exception to stdout?

File: src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs [hunk 8]
```diff
@@ -178,9 +188,19 @@ private static async Task CheckExceptionForBadConnectionStringHelperAsync(string
...
+                {
+                    Assert.AreEqual(HttpStatusCode.InternalServerError, ex.StatusCode);
```
Comment (Aniruddh25) [inline src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:197]: Why is this an internal server error? Why not serviceunavailable?
Did the server not come up due to invalid user input or something wrong with the server?

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
Good capture of all errors, few more suggestions.

File: src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs (hunk 8 above)
Comment (aaronburtle) [reply src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:197]: This is the exception that comes back trying to invoke `GetConfig()` through the `RuntimeConfigProvider` in the case where trying to load the known config results in that config not loading but instead being `null`. In this case we had decided on an `InternalServerError`. I think this makes sense because it is indicating back to the user that the problem is with our internal processing, in this case of the config file.

image

Comment (aaronburtle) [reply src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:198]: after discussion we decided to use `ServiceUnavailable` for all of the config related errors. this is the status code used in the config validator already, so changing this matches the rest of the messaging we use.

Comment (Aniruddh25) [reply src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs:198]: Since we decided to go with ServiceUnavailable, shouldn't the assert here be changed?

**Review by Aniruddh25 — APPROVED** [review]
LGTM, after fixing the remaining test failures. Thanks for the quick fix

File: src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs (hunk 8 above)
Comment (aaronburtle) [reply src/Service.Tests/Unittests/SqlMetadataProviderUnitTests.cs]: yes, I had this locally only, pushed the changes now.

**Review by seantleonard — APPROVED** [review]
thank you for getting this error handling improvement in and addressing feedback

**Review by seantleonard — COMMENTED** [review]
feedback
